### PR TITLE
Add popover bottom placements

### PR DESCRIPTION
**Pull request by phildobsontw**
Any popover that originates within 125 pixels of the left or right hand side of a 360px width viewport will overflow fairly significantly.

Old: (360px width viewport)
image

The fix: 
Adding two new bottom placements for the popover, BOTTOM_RIGHT and BOTTOM_LEFT.

These will be automatically selected if:
- Preferred position is left or right
- Placing the popover on both the left and the right would result in overflow
- Placing the popover below results in an overflow on one side (it will now therefore be shifted either left, or right).

Slight bit of refactoring to enable reuse of methods (and remove duplicate calculations).

New:
image

This isn't perfect - It's possible that with this logic up to 19px of the popover will overflow the left or right hand side. As the margin on the popover is 24px, this is a significant improvement as no text will be obscured.

### CODE REVIEW

File: src/popover/positioning.js [hunk 1]
```diff
@@ -1,46 +1,60 @@
...
+  } else if (overflowsWithBottomPlacement.right && !overflowsWithBottomPlacement.left) {
+    return BOTTOM_LEFT;
+  }
+  return BOTTOM;
+}
+
+function calculateOverflows(popover, placement, documentBounds) {
+  const pos = getPopoverPosition(popover, placement);
+  return {
+    right: pos.right > documentBounds.right,
```
Comment (andreapiras) [inline src/popover/positioning.js]: Maybe let's call it 'isOverflowRight' ?

File: src/popover/positioning.js [hunk 2]
```diff
@@ -1,46 +1,60 @@
...
+  if (!rightPlacementOverflows.right && !leftPlacementOverflows.left) {
+    return preferredPlacement;
   }
-  if (overflowsRight && overflowsLeft) {
-    return BOTTOM;
+  if (rightPlacementOverflows.right && leftPlacementOverflows.left) {
+    return determineBestBottomPlacement(popover, documentBounds);
   }
-  if (overflowsRight && placement === RIGHT) {
+  if (rightPlacementOverflows.right && preferredPlacement === RIGHT) {
```
Comment (andreapiras) [inline src/popover/positioning.js]: What if it does overflow left too?

File: src/popover/positioning.js [hunk 3]
```diff
@@ -1,46 +1,60 @@
...
     return RIGHT_TOP;
   }
-  return placement;
+  return preferredPlacement;
+}
+
+function determineBestBottomPlacement(popover, documentBounds) {
+  const overflowsWithBottomPlacement = calculateOverflows(popover, BOTTOM, documentBounds);
+  if (overflowsWithBottomPlacement.left && !overflowsWithBottomPlacement.right) {
+    return BOTTOM_RIGHT;
```
Comment (andreapiras) [inline src/popover/positioning.js]: Do we really need BOTTOM_LEFT AND BOTTOM_RIGHT?

File: src/popover/positioning.js (hunk 2 above)
Comment (andreapiras) [inline src/popover/positioning.js]: How do You know there s no overflow left

Comment (phildobsontw) [reply src/popover/positioning.js]: I think we'd only be concerned with whether the popover if positioned on the left would overflow on the left if we were to actually intend to position it on the left. This should be covered on line 33 👍

File: src/popover/positioning.js (hunk 3 above)
Comment (phildobsontw) [reply src/popover/positioning.js]: `bottom-left` and `bottom-right` are classes both already available in our bootstrap CSS. Having both of these enables us to shift the popover either left or right based on which side it overflows when positioned at the bottom.

File: src/popover/positioning.js (hunk 1 above)
Comment (phildobsontw) [reply src/popover/positioning.js]: have gone with `overflowsRight` just to avoid it sounding like a method (might just be my sketchy java background 😉).

**Pull request merged**